packages/core/src/utils/extensionLoader.ts

```diff
@@ -4,45 +4,194 @@
+    } else {
+      throw new Error('Already started, you may only call `start` once.');
+    }
+    await Promise.all(
+      this.getExtensions()
+        .filter((e) => e.isActive)
+        .map(this.startExtension.bind(this)),
+    );
+  }
+
+  /**
+   * Unconditionally starts an `extension` and loads all its MCP servers,
+   * context, custom commands, etc. Assumes that `start` has already been called
+   * and we have a Config object.
+   *
+   * This should typically only be called from `start`, most other calls should
+   * go through `maybeStartExtension` which will only start the extension if
+   * extension reloading is enabled and the `config` object is initialized.
+   */
+  protected async startExtension(extension: GeminiCLIExtension) {
+    if (!this.config) {
+      throw new Error('Cannot call `startExtension` prior to calling `start`.');
+    }
+    this.startingCount++;
+    this.eventEmitter?.emit('extensionsStarting', {
+      total: this.startingCount,
+      completed: this.startCompletedCount,
+    });
+    try {
+      await this.config.getMcpClientManager()!.startExtension(extension);
+      // TODO: Move all extension features here, including at least:
+      // - context file loading
+      // - custom command loading
+      // - excluded tool configuration
+    } finally {
+      this.startCompletedCount++;
+      this.eventEmitter?.emit('extensionsStarting', {
+        total: this.startingCount,
+        completed: this.startCompletedCount,
+      });
+      if (this.startingCount === this.startCompletedCount) {
+        this.startingCount = 0;
+        this.startCompletedCount = 0;
+      }
+    }
+  }
+
+  /**
+   * If extension reloading is enabled and `start` has already been called,
+   * then calls `startExtension` to include all extension features into the
+   * program.
+   */
+  protected maybeStartExtension(
+    extension: GeminiCLIExtension,
+  ): Promise<void> | undefined {
+    if (this.config && this.config.getEnableExtensionReloading()) {
+      return this.startExtension(extension);
+    }
+    return;
+  }
+
+  /**
+   * Unconditionally stops an `extension` and unloads all its MCP servers,
+   * context, custom commands, etc. Assumes that `start` has already been called
+   * and we have a Config object.
+   *
+   * Most calls should go through `maybeStopExtension` which will only stop the
+   * extension if extension reloading is enabled and the `config` object is
+   * initialized.
+   */
+  protected async stopExtension(extension: GeminiCLIExtension) {
+    if (!this.config) {
+      throw new Error('Cannot call `stopExtension` prior to calling `start`.');
+    }
+    this.stoppingCount++;
+    this.eventEmitter?.emit('extensionsStopping', {
+      total: this.stoppingCount,
+      completed: this.stopCompletedCount,
+    });
+
+    try {
+      await this.config.getMcpClientManager()!.stopExtension(extension);
+      // TODO: Remove all extension features here, including at least:
```

Comment:
can you resolve this todo?

Reply:
This will be in a follow-up - this entire feature is guarded behind an experiment right now until these are all addressed.